source/val/validate_memory.cpp

```diff
@@ -180,6 +180,96 @@ bool ContainsInvalidBool(ValidationState_t& _, const Instruction* storage,
   return false;
 }
 
+std::pair<SpvStorageClass, SpvStorageClass> GetStorageClass(
+    ValidationState_t& _, const Instruction* inst) {
+  SpvStorageClass dst_sc = SpvStorageClassMax;
+  SpvStorageClass src_sc = SpvStorageClassMax;
+  switch (inst->opcode()) {
+    case SpvOpLoad: {
+      auto load_pointer = _.FindDef(inst->GetOperandAs<uint32_t>(2));
+      auto load_pointer_type = _.FindDef(load_pointer->type_id());
+      dst_sc = load_pointer_type->GetOperandAs<SpvStorageClass>(1);
+      break;
+    }
+    case SpvOpStore: {
+      auto store_pointer = _.FindDef(inst->GetOperandAs<uint32_t>(0));
+      auto store_pointer_type = _.FindDef(store_pointer->type_id());
+      dst_sc = store_pointer_type->GetOperandAs<SpvStorageClass>(1);
+      break;
+    }
+    case SpvOpCopyMemory:
+    case SpvOpCopyMemorySized: {
+      auto dst = _.FindDef(inst->GetOperandAs<uint32_t>(0));
+      auto dst_type = _.FindDef(dst->type_id());
+      dst_sc = dst_type->GetOperandAs<SpvStorageClass>(1);
+      auto src = _.FindDef(inst->GetOperandAs<uint32_t>(1));
+      auto src_type = _.FindDef(src->type_id());
+      src_sc = src_type->GetOperandAs<SpvStorageClass>(1);
+      break;
+    }
+    default:
+      break;
+  }
+
+  return std::make_pair(dst_sc, src_sc);
+}
+
+spv_result_t CheckMemoryAccess(ValidationState_t& _, const Instruction* inst,
+                               uint32_t mask) {
+  if (mask & SpvMemoryAccessMakePointerAvailableKHRMask) {
+    if (inst->opcode() == SpvOpLoad) {
+      return _.diag(SPV_ERROR_INVALID_ID, inst)
+             << "MakePointerAvailableKHR cannot be used with OpLoad.";
+    }
+
+    if (!(mask & SpvMemoryAccessNonPrivatePointerKHRMask)) {
+      return _.diag(SPV_ERROR_INVALID_ID, inst)
+             << "NonPrivatePointerKHR must be specified if "
+                "MakePointerAvailableKHR is specified.";
+    }
+  }
+
+  if (mask & SpvMemoryAccessMakePointerVisibleKHRMask) {
+    if (inst->opcode() == SpvOpStore) {
+      return _.diag(SPV_ERROR_INVALID_ID, inst)
+             << "MakePointerVisibleKHR cannot be used with OpStore.";
+    }
+
+    if (!(mask & SpvMemoryAccessNonPrivatePointerKHRMask)) {
+      return _.diag(SPV_ERROR_INVALID_ID, inst)
+             << "NonPrivatePointerKHR must be specified if "
+                "MakePointerVisibleKHR is specified.";
+    }
+  }
+
+  if (mask & SpvMemoryAccessNonPrivatePointerKHRMask) {
+    SpvStorageClass dst_sc, src_sc;
+    std::tie(dst_sc, src_sc) = GetStorageClass(_, inst);
+    if (dst_sc != SpvStorageClassUniform &&
```

Comment:
Weird asymmetry here:  this dst_sc does not check SpvStorageClassMax but the src_sc does.

Reply:
All instructions use dst_sc (maybe a bad name), but only the copies use src_sc. So the condition is whether or not src_sc  should be a check.